src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -1684,6 +1685,299 @@ private int GetLastWin32ErrorAndDisposeHandleIfInvalid(bool throwIfInvalidHandle
+            return _isAsync ?
+                AsyncModeCopyToAsync(destination, bufferSize, cancellationToken) :
+                base.CopyToAsync(destination, bufferSize, cancellationToken);
+        }
+
+        private async Task AsyncModeCopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+        {
+            Debug.Assert(_isAsync, "This implementation is for async mode only");
+            Debug.Assert(!_handle.IsClosed, "!_handle.IsClosed");
+            Debug.Assert(_parent.CanRead, "_parent.CanRead");
+
+            // Make sure any pending writes have been flushed before we do a read.
+            if (_writePos > 0)
+            {
+                await FlushWriteAsync(cancellationToken).ConfigureAwait(false);
+            }
+
+            // Typically CopyToAsync would be invoked as the only "read" on the stream, but it's possible some reading is
+            // done and then the CopyToAsync is issued.  For that case, see if we have any data available in the buffer.
+            if (_buffer != null)
+            {
+                int bufferedBytes = _readLen - _readPos;
+                if (bufferedBytes > 0)
+                {
+                    await destination.WriteAsync(_buffer, _readPos, bufferedBytes, cancellationToken).ConfigureAwait(false);
+                    _readPos = _readLen = 0;
+                }
+            }
+
+            // For efficiency, we avoid creating a new task and associated state for each asynchronous read.
+            // Instead, we create a single reusable awaitable object that will be triggered when an await completes
+            // and reset before going again.
+            var readAwaitable = new AsyncCopyToAwaitable(this);
+
+            // Make sure we are reading from the position that we think we are.
+            // Only set the position in the awaitable if we can seek (e.g. not for pipes).
+            bool canSeek = _parent.CanSeek;
+            if (canSeek)
+            {
+                if (_exposedHandle) VerifyOSHandlePosition();
+                readAwaitable._position = _pos;
+            }
+
+            // Create the buffer to use for the copy operation, as the base CopyToAsync does. We don't try to use
+            // _buffer here, even if it's not null, as concurrent operations are allowed, and another operation may
+            // actually be using the buffer already. Plus, it'll be rare for _buffer to be non-null, as typically
+            // CopyToAsync is used as the only operation performed on the stream, and the buffer is lazily initialized.
+            // Further, typically the CopyToAsync buffer size will be larger than that used by the FileStream, such that
+            // we'd likely be unable to use it anyway.  A better option than using _buffer would be a future pooling solution.
+            byte[] copyBuffer = new byte[bufferSize];
```

Comment:
I would like to.  I previously tried that in CopyToAsync, though, and we backed it out.
https://github.com/dotnet/corefx/pull/5954
https://github.com/dotnet/corefx/pull/5954#discussion_r52130236

I'm reading through all the comments again, though, and there wasn't really a strong conclusion about using it in CopyToAsync specifically (or if there was, I'm missing it).  The danger in a case like this is that we pass the buffer to the target stream's WriteAsync, and its WriteAsync holds on to the buffer beyond when the task it returns completes.  Then we return the buffer to the pool, and some other code elsewhere uses the same buffer, while that WriteAsync may still be erroneously using it.

But, this seems like a reasonable risk to me for the benefit pooling would provide.  These are typically very large buffers, and it'd be great to avoid re-allocating it for each copy.  Especially since with an optimization like this, we're now forcing users to choose between the benefits of the optimized CopyToAsync from FileStream, and using their own copying implementation that does pool a buffer and just reads/writes from source to target.  The main concern with pooling in our streams is non-localized corruption, something over here causing a problem with something over there that's completely unrelated... a valid concern, but the liklihood of that in this situation is low and highlights a very buggy stream implementation that's likely to manifest problems in other ways as well (if the task is returned before it's actually done its work).

@jkotas, @KrzysztofCwalina, @ericstj, @weshaggard, could you guys weigh in again on this specific case, using ```ArrayPool<byte>.Default``` to get a buffer used in our stream's CopyToAsync overrides?